### PR TITLE
graphite: 1.0.2 -> 1.1.3 & moved dependencies to python-modules

### LINKED ISSUE #30891: Graphite-related packages should move out of pythonPackages
## Issue description
I'm cleaning up `pythonPackages` and noticed that some dependencies of some `graphite` dependencies (`txamqp`) are several years out of date.

Since graphite is not a library it should move out of `pythonPackages` into it's own package. 

cc @rickynils @FRidh

### PR DESCRIPTION
**Pull request by basvandijk**
###### Motivation for this change

Fixes #30891 and closes https://github.com/NixOS/nixpkgs/pull/44246.

###### Things done

* Upgrade `graphite-web`, `carbon` and `whisper` from 1.0.2 -> 1.1.3.

* Extend the `pythonPackages.callPackage` scope with the
  `sharedVersions` attribute set. This set contains versions that are
  shared by multiple python packages. Currently it only contains the
  `graphiteVersion`.

* Replaced the deprecated `pythonPackages.graphite_influxdb` with
  `pythonPackages.influxgraph.`

* Renamed `pythonPackages.graphite_web` to `pythonPackages.graphite-web`
  to be consistent with the Python package name.

* Replaced the unmaintained `pythonPackages.graphite_pager` with
  `pythonPackages.graphitepager`

* Moved all new packages from `python-packages.nix` to
  `pkgs/development/python-modules`

* The following test passes:
  `nix-build nixos/release.nix -A tests.graphite.x86_64-linux`

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [x] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [x] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

@dotlambda @FRidh @adisbladis @offlinehacker @rickynils

### CODE REVIEW

**Comment by dotlambda** [comment]
What's the advantage of having `graphiteVersion` instead of specifying the version for each package? Do they really need to be the same version? E.g., `graphite_api` seems to work even though it is at a different version.
I'm asking because specifying the version individually would allow us to include the graphite packages in batch upgrades.

**Comment by Mic92** [comment]
@GrahamcOfBorg test graphite

**Comment by basvandijk** [comment]
@dotlambda it could be possible that different versions of `graphite-web`, `carbon` and `whisper` can work together. However, since they share the same version, are published on the same date and [share the same release notes](https://graphite.readthedocs.io/en/latest/releases/1_1_3.html) I think upstream intends them to be used together. Having a shared `graphiteVersion` ensures this in `nixpkgs`. 

Note that we do [something similar with the Elasticsearch stack](https://github.com/NixOS/nixpkgs/blob/e8d6ceaf397959250b56fc5600144f9e25840b71/pkgs/top-level/all-packages.nix#L2263).

Having support for batch upgrades would also be nice of course. I guess batch upgrades can also ensure we upgrade all graphite packages together. What needs to be done to get the graphite packages to batch upgrade?

File: pkgs/top-level/python-packages.nix [hunk 1]
```diff
@@ -29,7 +29,14 @@ let
   isPyPy = python.executable == "pypy";
   isPy3k = strings.substring 0 1 python.pythonVersion == "3";
 
-  callPackage = pkgs.newScope self;
+  callPackage = pkgs.newScope (self // sharedVersions);
+
+  # Some Python packages are released together and share the same version.
+  # These versions are collected in the sharedVersions attribute set and are
+  # added to the callPackage scope so that packages can refer to these versions.
+  sharedVersions = {
```
Comment (FRidh) [inline pkgs/top-level/python-packages.nix]: I like the idea, but as there currently is only one version I don't think we should have this. If we were to have this, I think this should go to top-level so we could also use it with say `mercurial` and `tortoisehg` and other cases there.

Comment (dotlambda) [reply pkgs/top-level/python-packages.nix]: Another example: `libgit2` and `pygit2`.

**Comment by dotlambda** [comment]
> What needs to be done to get the graphite packages to batch upgrade?

Currently, our script can only deal with the version being specified in each file. If new versions are published on the same date, they would be updated simultaneously anyway.

**Comment by dotlambda** [comment]
There's no need to add packages manually. It's applied to all packages in `pkgs/development/python-modules`. You'll find it here: https://github.com/NixOS/nixpkgs/blob/master/maintainers/scripts/update-python-libraries

**Comment by basvandijk** [comment]
Great. I've updated my PR and removed `graphiteVersion`.

File: pkgs/development/python-modules/graphite-web/default.nix [hunk 2]
```diff
@@ -0,0 +1,55 @@
+{ stdenv, buildPythonPackage, fetchPypi, isPy3k, which
+, django, django_1_8
+, django_tagging, django_tagging_0_4_3
+, whisper, pycairo, cairocffi, ldap, memcached, pytz, urllib3, scandir
+}:
+if django         != django_1_8
+|| django_tagging != django_tagging_0_4_3
```
Comment (dotlambda) [inline pkgs/development/python-modules/graphite-web/default.nix]: Can we instead use something like `django.version != "1.8"`?

File: pkgs/development/python-modules/graphite-web/default.nix [hunk 3]
```diff
@@ -0,0 +1,55 @@
...
+  };
+
+  propagatedBuildInputs = [
+    django django_tagging whisper pycairo cairocffi
+    ldap memcached pytz urllib3 scandir
+  ];
+
+  postInstall = ''
+    wrapProgram $out/bin/run-graphite-devel-server.py \
+      --prefix PATH : ${which}/bin
```
Comment (dotlambda) [inline pkgs/development/python-modules/graphite-web/default.nix]: You can use `makeWrapperArgs` instead.

File: pkgs/development/python-modules/graphite_beacon/default.nix [hunk 4]
```diff
@@ -0,0 +1,23 @@
...
+  version = "0.27.0";
+
+  src = fetchPypi {
+    inherit pname version;
+    sha256 = "03bp4wyfn3xhcqyvs5hnk1n87m4smsmm1p7qp459m7j8hwpbq2ks";
+  };
+
+  propagatedBuildInputs = [ tornado pyyaml funcparserlib ];
+
+  patchPhase = "> requirements.txt";
```
Comment (dotlambda) [inline pkgs/development/python-modules/graphite_beacon/default.nix]: Why is this needed?

File: pkgs/development/python-modules/graphitepager/default.nix [hunk 5]
```diff
@@ -0,0 +1,30 @@
...
+    inherit pname version;
+    sha256 = "0v3g1qcgnkpgjzh6phnv13lnk8qjrcs9sq2qg6k0dk5ik31jfk3d";
+  };
+
+  propagatedBuildInputs = [
+    jinja2 markupsafe pagerduty pushbullet python_magic python-simple-hipchat
+    pyyaml redis requests six websocket_client
+  ];
+
+  patchPhase = "> requirements.txt";
```
Comment (dotlambda) [inline pkgs/development/python-modules/graphitepager/default.nix]: Here as well. Would be best to get rid of it.

File: pkgs/development/python-modules/graphite-web/default.nix (hunk 3 above)
Comment (basvandijk) [reply pkgs/development/python-modules/graphite-web/default.nix]: I tried the following:
```
makeWrapperArgs = [ "--prefix PATH : ${which}/bin" ];
```
but with that the test suite fails to start up graphite-web:

```
one# [    7.566444] graphiteWeb-pre-start[706]: Traceback (most recent call last):
one# [    7.568265] graphiteWeb-pre-start[706]:   File "/nix/store/k8xndnycd95iqs0ky9nlcvqqnyx9p63v-python2.7-Django-1.8.18/bin/..django-admin.py-wrapped-wrapped", line 7, in <module>
one# [    7.574263] graphiteWeb-pre-start[706]:     management.execute_from_command_line()
one# [    7.582220] graphiteWeb-pre-start[706]:   File "/nix/store/h5pfcf6j3rl85xm4apggqg1d4za4yj53-python-2.7.15-env/lib/python2.7/site-packages/django/core/management/__init__.py", line 354, in execute_from_command_line
one# [    7.595708] graphiteWeb-pre-start[706]:     utility.execute()
one# [    7.602375] graphiteWeb-pre-start[706]:   File "/nix/store/h5pfcf6j3rl85xm4apggqg1d4za4yj53-python-2.7.15-env/lib/python2.7/site-packages/django/core/management/__init__.py", line 328, in execute
one# [    7.611485] graphiteWeb-pre-start[706]:     django.setup()
one# [    7.617331] graphiteWeb-pre-start[706]:   File "/nix/store/h5pfcf6j3rl85xm4apggqg1d4za4yj53-python-2.7.15-env/lib/python2.7/site-packages/django/__init__.py", line 18, in setup
one# [    7.626695] graphiteWeb-pre-start[706]:     apps.populate(settings.INSTALLED_APPS)
one# [    7.635331] graphiteWeb-pre-start[706]:   File "/nix/store/h5pfcf6j3rl85xm4apggqg1d4za4yj53-python-2.7.15-env/lib/python2.7/site-packages/django/apps/registry.py", line 85, in populate
one# [    7.646523] graphiteWeb-pre-start[706]:     app_config = AppConfig.create(entry)
one# [    7.652897] graphiteWeb-pre-start[706]:   File "/nix/store/h5pfcf6j3rl85xm4apggqg1d4za4yj53-python-2.7.15-env/lib/python2.7/site-packages/django/apps/config.py", line 86, in create
one# [    7.665890] graphiteWeb-pre-start[706]:     module = import_module(entry)
one# [    7.674238] graphiteWeb-pre-start[706]:   File "/nix/store/7sbvn0wgv7hsnxpss1jba723kx5nz6d4-python-2.7.15/lib/python2.7/importlib/__init__.py", line 37, in import_module
one# [    7.682407] graphiteWeb-pre-start[706]:     __import__(name)
one# [    7.693954] graphiteWeb-pre-start[706]:   File "/nix/store/h5pfcf6j3rl85xm4apggqg1d4za4yj53-python-2.7.15-env/lib/python2.7/site-packages/opt/graphite/webapp/graphite/functions/__init__.py", line 10, in <module>
one# [    7.703974] graphiteWeb-pre-start[706]:     from graphite.logger import log
one# [    7.710845] graphiteWeb-pre-start[706]:   File "/nix/store/h5pfcf6j3rl85xm4apggqg1d4za4yj53-python-2.7.15-env/lib/python2.7/site-packages/opt/graphite/webapp/graphite/logger.py", line 106, in <module>
one# [    7.720270] graphiteWeb-pre-start[706]:     log = GraphiteLogger() # import-shared logger instance
one# [    7.725289] graphiteWeb-pre-start[706]:   File "/nix/store/h5pfcf6j3rl85xm4apggqg1d4za4yj53-python-2.7.15-env/lib/python2.7/site-packages/opt/graphite/webapp/graphite/logger.py", line 39, in __init__
one# [    7.732729] graphiteWeb-pre-start[706]:     level = logging.DEBUG if settings.DEBUG else logging.INFO,
one# [    7.736779] graphiteWeb-pre-start[706]:   File "/nix/store/h5pfcf6j3rl85xm4apggqg1d4za4yj53-python-2.7.15-env/lib/python2.7/site-packages/opt/graphite/webapp/graphite/logger.py", line 78, in _config_logger
one# [    7.743253] graphiteWeb-pre-start[706]:     handler = Rotater(log_file, when=when, backupCount=backupCount)
one# [    7.748248] graphiteWeb-pre-start[706]:   File "/nix/store/7sbvn0wgv7hsnxpss1jba723kx5nz6d4-python-2.7.15/lib/python2.7/logging/handlers.py", line 171, in __init__
one# [    7.754796] graphiteWeb-pre-start[706]:     BaseRotatingHandler.__init__(self, filename, 'a', encoding, delay)
one# [    7.762192] graphiteWeb-pre-start[706]:   File "/nix/store/7sbvn0wgv7hsnxpss1jba723kx5nz6d4-python-2.7.15/lib/python2.7/logging/handlers.py", line 64, in __init__
one# [    7.767696] graphiteWeb-pre-start[706]:     logging.FileHandler.__init__(self, filename, mode, encoding, delay)
one# [    7.771913] graphiteWeb-pre-start[706]:   File "/nix/store/7sbvn0wgv7hsnxpss1jba723kx5nz6d4-python-2.7.15/lib/python2.7/logging/__init__.py", line 920, in __init__
one# [    7.779395] graphiteWeb-pre-start[706]:     StreamHandler.__init__(self, self._open())
one# [    7.784128] graphiteWeb-pre-start[706]:   File "/nix/store/7sbvn0wgv7hsnxpss1jba723kx5nz6d4-python-2.7.15/lib/python2.7/logging/__init__.py", line 950, in _open
one# [    7.789363] graphiteWeb-pre-start[706]:     stream = open(self.baseFilename, self.mode)
one# [    7.794295] graphiteWeb-pre-start[706]: IOError: [Errno 2] No such file or directory: '/var/db/graphite/log/webapp/info.log'
```

File: pkgs/development/python-modules/graphite_beacon/default.nix (hunk 4 above)
Comment (basvandijk) [reply pkgs/development/python-modules/graphite_beacon/default.nix]: This was in the [original package](https://github.com/NixOS/nixpkgs/pull/44276/files#diff-2afbba83e1b14f9e0c304c769831a0e2L15670) and without it the build fails at the dependency checking phase. I do feel a bit uneasy about just ignoring dependency version ranges.

Comment (dotlambda) [reply pkgs/development/python-modules/graphite_beacon/default.nix]: How about using
```
substituteInPlace requirements.txt "==" ">="
```
instead?
Also, this should be put into `postPatch`.

Comment (basvandijk) [reply pkgs/development/python-modules/graphite_beacon/default.nix]: Thanks, that's better indeed.

**Comment by basvandijk** [comment]
@dotlambda @FRidh I think this is ready to be merged now.

I also performed a manual test by modifying the graphite test to enable an xserver and install firefox. Then I ran the VM using `$(nix-build nixos/tests/graphite.nix -A driver)/bin/nixos-run-vms` and pointed firefox to graphite-web on localhost:8080. I could successfully operate the app.

**Comment by FRidh** [comment]
Unrelated, but:
```
  carbonEnv = {
    PYTHONPATH = let
      cenv = pkgs.python.buildEnv.override {
        extraLibs = [ pkgs.python27Packages.carbon ];
      };
      cenvPack =  "${cenv}/${pkgs.python.sitePackages}";
    # opt/graphite/lib contains twisted.plugins.carbon-cache
    in "${cenvPack}/opt/graphite/lib:${cenvPack}";
```
How about moving `opt/graphite/lib` to a Python package?

**Review by FRidh — APPROVED** [review]

File: pkgs/development/python-modules/graphite-web/default.nix [hunk 6]
```diff
@@ -0,0 +1,53 @@
...
+  '';
+
+  # error: invalid command 'test'
+  doCheck = false;
+
+  meta = {
+    homepage = http://graphite.wikidot.com/;
+    description = "Enterprise scalable realtime graphing";
+    maintainers = with stdenv.lib.maintainers; [ rickynils offline basvandijk ];
+  };
```
Comment (dotlambda) [inline pkgs/development/python-modules/graphite-web/default.nix]: `license`

File: pkgs/development/python-modules/influxgraph/default.nix [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+
+  propagatedBuildInputs = [ influxdb graphite_api memcached ];
+
+  passthru.moduleName = "influxgraph.InfluxDBFinder";
+
+  meta = with stdenv.lib; {
+    description = "InfluxDB storage plugin for Graphite-API";
+    homepage = https://github.com/InfluxGraph/influxgraph;
+    license = licenses.asl20;
+    platforms = platforms.linux;
```
Comment (dotlambda) [inline pkgs/development/python-modules/influxgraph/default.nix]: I think this can be dropped. Or is it needed?

File: pkgs/development/python-modules/whisper/default.nix [hunk 8]
```diff
@@ -0,0 +1,19 @@
...
+    sha256 = "1ahzsxk52ws8k3kdq52qbsbsx2r9z350j8gg9adw4x5fjwksz4r8";
+  };
+
+  propagatedBuildInputs = [ six ];
+
+  meta = {
+    homepage = http://graphite.wikidot.com/;
+    description = "Fixed size round-robin style database";
+    maintainers = with stdenv.lib.maintainers; [ rickynils offline basvandijk ];
+  };
```
Comment (dotlambda) [inline pkgs/development/python-modules/whisper/default.nix]: `license`

File: pkgs/development/python-modules/influxgraph/default.nix (hunk 7 above)
Comment (basvandijk) [reply pkgs/development/python-modules/influxgraph/default.nix]: Yes, I think this can be dropped. I'll do it in a sec.

**Comment by dotlambda** [comment]
Btw, should the changes to the module be mentioned in the changelog?

**Comment by basvandijk** [comment]
@dotlambda thanks for the merge. The API of the graphite module hasn't changed. So existing invocations of the module don't have to be changed. So I don't think a changelog entry is needed.

**Comment by basvandijk** [comment]
@FRidh about making `$(nix-build -A python27Packages.carbon)/lib/python2.7/site-packages/opt/graphite/lib/` a Python package: this should ideally be done by upstream right?
